packages/devtools_app/lib/src/extensions/extension_service.dart

```diff
@@ -14,17 +15,36 @@ class ExtensionService extends DisposableController
       _availableExtensions;
   final _availableExtensions = ValueNotifier<List<DevToolsExtensionConfig>>([]);
 
-  void initialize() {
-    addAutoDisposeListener(serviceManager.connectedState, () {
-      _refreshAvailableExtensions();
+  Future<void> initialize() async {
+    await maybeRefreshExtensions();
+    addAutoDisposeListener(serviceManager.connectedState, () async {
+      await maybeRefreshExtensions();
     });
+
+    // TODO(kenz): we should also refresh the available extensions on some event
+    // from the analysis server that is watching the
+    // .dart_tool/package_config.json file for changes.
+  }
+
+  Future<void> maybeRefreshExtensions() async {
+    final appRootPath = await _connectedAppRootPath();
+    if (appRootPath != null) {
+      await _refreshAvailableExtensions(appRootPath);
+    }
   }
 
-  // TODO(kenz): actually look up the available extensions from devtools server,
-  // based on the root path(s) from the available isolate(s).
-  int _count = 0;
-  void _refreshAvailableExtensions() {
-    _availableExtensions.value =
-        debugExtensions.sublist(0, _count++ % (debugExtensions.length + 1));
+  Future<void> _refreshAvailableExtensions(String? rootPath) async {
+    final extensions = await server.refreshAvailableExtensions(rootPath);
+    _availableExtensions.value = extensions;
+  }
+}
+
+Future<String?> _connectedAppRootPath() async {
+  var fileUri = await serviceManager.rootLibraryForSelectedIsolate();
+  if (fileUri == null) return null;
+
+  if (fileUri.endsWith('/lib/main.dart')) {
```

Comment:
What happens if the root library isn't `main.dart`? There's no requirement for the entry point of a Dart or Flutter program to be in `main.dart`. I'm not even sure there's a requirement for it to be in a library of the form `lib/<script>` for Flutter, but there definitely isn't for Dart programs.

Reply:
Is there a convention for finding the root directory that contains the `.dart_tool` folder? Is this guaranteed to always be the parent of `lib/`

Reply:
I'm not sure, especially if you don't have access to the FS. If you did, you'd just walk the directory chain until you found the directory containing `.dart_tool`.

I think you're probably safe to assume that the entrypoint is under `lib/`, so if you update this to `fileUri.endsWith(RegExp(r'\/lib\/[^\/.]*.dart'))` this should work the vast majority of the time.

Reply:
Optionally, I could just pass the whole root library to the server and have the server look for the .dart_tool folder?

Reply:
That'd be the most accurate way of handling it.

Reply:
Used the regexp for now but added a todo to move this logic to the server.